src/MOI_wrapper.jl

```diff
@@ -0,0 +1,1422 @@
+quadratic_le_offset(model::Optimizer) = linear_eq_offset(model) + length(model.linear_eq_constraints)
+quadratic_ge_offset(model::Optimizer) = quadratic_le_offset(model) + length(model.quadratic_le_constraints)
+quadratic_eq_offset(model::Optimizer) = quadratic_ge_offset(model) + length(model.quadratic_ge_constraints)
+nlp_constraint_offset(model::Optimizer) = quadratic_eq_offset(model) + length(model.quadratic_eq_constraints)
+
+# Convenience functions used only in optimize!
+
+function append_to_jacobian_sparsity!(jacobian_sparsity, aff::MOI.ScalarAffineFunction, row)
+    for term in aff.terms
+        push!(jacobian_sparsity, (row, term.variable_index.value))
+    end
+end
+
+function append_to_jacobian_sparsity!(jacobian_sparsity, quad::MOI.ScalarQuadraticFunction, row)
+    for term in quad.affine_terms
+        push!(jacobian_sparsity, (row, term.variable_index.value))
+    end
+    for term in quad.quadratic_terms
+        row_idx = term.variable_index_1
+        col_idx = term.variable_index_2
+        if row_idx == col_idx
+            push!(jacobian_sparsity, (row, row_idx.value))
+        else
+            push!(jacobian_sparsity, (row, row_idx.value))
+            push!(jacobian_sparsity, (row, col_idx.value))
+        end
+    end
+end
+
+# Refers to local variables in jacobian_structure() below.
+macro append_to_jacobian_sparsity(array_name)
+    escrow = esc(:row)
+    quote
+        for info in $(esc(array_name))
+            append_to_jacobian_sparsity!($(esc(:jacobian_sparsity)), info.func, $escrow)
+            $escrow += 1
+        end
+    end
+end
+
+function jacobian_structure(model::Optimizer)
+    num_nlp_constraints = length(model.nlp_data.constraint_bounds)
+    if num_nlp_constraints > 0
+        nlp_jacobian_sparsity = MOI.jacobian_structure(model.nlp_data.evaluator)
+    else
+        nlp_jacobian_sparsity = []
+    end
+
+    jacobian_sparsity = Tuple{Int64,Int64}[]
+    row = 1
+    @append_to_jacobian_sparsity model.linear_le_constraints
+    @append_to_jacobian_sparsity model.linear_ge_constraints
+    @append_to_jacobian_sparsity model.linear_eq_constraints
+    @append_to_jacobian_sparsity model.quadratic_le_constraints
+    @append_to_jacobian_sparsity model.quadratic_ge_constraints
+    @append_to_jacobian_sparsity model.quadratic_eq_constraints
+    for (nlp_row, column) in nlp_jacobian_sparsity
+        push!(jacobian_sparsity, (nlp_row + row - 1, column))
+    end
+    return jacobian_sparsity
+end
+
+append_to_hessian_sparsity!(hessian_sparsity, ::Union{MOI.SingleVariable,MOI.ScalarAffineFunction}) = nothing
+
+function append_to_hessian_sparsity!(hessian_sparsity, quad::MOI.ScalarQuadraticFunction)
+    for term in quad.quadratic_terms
+        push!(hessian_sparsity, (term.variable_index_1.value,
+                                 term.variable_index_2.value))
+    end
+end
+
+function hessian_lagrangian_structure(model::Optimizer)
```

Comment:
IIRC NLopt doesn't take hessians, so all of this code can be deleted.

Reply:
Indeed, this is now removed.